sns_ik_lib/src/sns_acceleration_ik.cpp

```diff
@@ -0,0 +1,190 @@
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <sns_ik/sns_acceleration_ik.hpp>
+
+#include <ros/console.h>
+
+#include "sns_ik_math_utils.hpp"
+
+namespace sns_ik {
+
+static const Eigen::IOFormat EigArrFmt(4, 0, ", ", "\n", "[", "]");
+
+SNSAccelerationIK::SNSAccelerationIK(int dof, double loop_period) :
+  n_dof(0),
+  n_tasks(0),
+  m_usePositionLimits(true)
+{
+  setNumberOfDOF(dof);
+  setLoopPeriod(loop_period);
+
+  baseIkSolver = SnsAccIkBase::create(n_dof);
+}
+
+void SNSAccelerationIK::setNumberOfDOF(int dof)
+{
+  if (dof > 0 && dof != n_dof) {
+    n_dof = dof;
+  }
+}
+
+void SNSAccelerationIK::setNumberOfTasks(int ntasks, int dof)
+{
+  setNumberOfDOF(dof);
+
+  if (n_tasks != ntasks) {
+    n_tasks = ntasks;
+    Eigen::VectorXd ddq = Eigen::VectorXd::Zero(n_dof);
+
+    double scale = 1.0;
+    scaleFactors.resize(n_tasks, scale);
+  }
+}
+
+
+bool SNSAccelerationIK::setJointsCapabilities(const Eigen::VectorXd limit_low, const Eigen::VectorXd limit_high,
+                                          const Eigen::VectorXd maxVelocity, const Eigen::VectorXd maxAcceleration)
+{
+  if (limit_low.rows() != n_dof || limit_high.rows() != n_dof
+      || maxVelocity.rows() != n_dof || maxAcceleration.rows() != n_dof) {
+    return false;
+  }
+
+  jointLimit_low = limit_low;
+  jointLimit_high = limit_high;
+  maxJointVelocity = maxVelocity;
+  maxJointAcceleration = maxAcceleration;
+
+  ddotQmin = -maxJointAcceleration.array();
+  ddotQmax = maxJointAcceleration.array();
+
+  return true;
+}
+
+bool SNSAccelerationIK::setMaxJointAcceleration(const Eigen::VectorXd maxAcceleration)
+{
+  if (maxAcceleration.rows() != n_dof) {
+    return false;
+  }
+
+  maxJointAcceleration = maxAcceleration;
+  ddotQmin = -maxAcceleration.array();
+  ddotQmax = maxAcceleration.array();
+
+  return true;
+}
+
+void SNSAccelerationIK::shapeJointAccelerationBound(const Eigen::VectorXd &actualJointConfiguration,
+                                        const Eigen::VectorXd &actualJointVelocities, double margin) {
+
+  // it could be written using the Eigen::Array potentiality
+  double step, max, stop;
+
+  for (int i = 0; i < n_dof; i++) {
+    // for the minimum bound
+    max = -maxJointAcceleration(i);
+    if (m_usePositionLimits) {
+      step = 2 * (jointLimit_low(i) - actualJointConfiguration(i) - actualJointVelocities(i)*loop_period) / (loop_period * loop_period);
```

Comment:
is this taken from something specific I'm not aware of? worth a comment?

Reply:
It is based on the equations in the original paper. I will add a comment.